java/src/main/native/src/row_conversion.cu

```diff
@@ -1050,136 +1020,102 @@ copy_validity_from_rows(const size_type num_rows, const size_type num_columns,
-                                        validity_offset + absolute_col / CHAR_BIT];
-
-        // so every thread that is participating in the warp has a byte, but it's row-based
-        // data and we need it in column-based. So we shuffle the bits around to make
-        // the bytes we actually write.
-        for (int i = 0, byte_mask = 1; i < CHAR_BIT && relative_col + i < num_columns;
-             ++i, byte_mask <<= 1) {
-          auto validity_data = __ballot_sync(participation_mask, my_byte & byte_mask);
-          // lead thread in each warp writes data
-          if (threadIdx.x % warp_size == 0) {
-            auto const validity_write_offset =
-                validity_data_col_length * (relative_col + i) + relative_row / CHAR_BIT;
-
-            *reinterpret_cast<int32_t *>(&this_shared_tile[validity_write_offset]) = validity_data;
-          }
+  auto const tile = tile_infos[blockIdx.x];
+  auto const tile_start_col = tile.start_col;
+  auto const tile_start_row = tile.start_row;
+  auto const num_tile_cols = tile.num_cols();
+  auto const num_tile_rows = tile.num_rows();
+
+  auto const threads_per_warp = warp.size();
+  auto const cols_per_read = CHAR_BIT;
+
+  auto const rows_per_read = static_cast<size_type>(threads_per_warp);
+  auto const num_sections_x = util::div_rounding_up_safe(num_tile_cols, cols_per_read);
+  auto const num_sections_y = util::div_rounding_up_safe(num_tile_rows, rows_per_read);
+  auto const validity_data_col_length = num_sections_y * 4; // words to bytes
+  auto const total_sections = num_sections_x * num_sections_y;
+
+  // the tile is divided into sections. A warp operates on a section at a time.
+  for (int my_section_idx = warp.meta_group_rank(); my_section_idx < total_sections;
+       my_section_idx += warp.meta_group_size()) {
+    // convert section to row and col
+    auto const section_x = my_section_idx % num_sections_x;
+    auto const section_y = my_section_idx / num_sections_x;
+    auto const relative_col = section_x * cols_per_read;
+    auto const relative_row = section_y * rows_per_read + warp.thread_rank();
+    auto const absolute_col = relative_col + tile_start_col;
+    auto const absolute_row = relative_row + tile_start_row;
+    auto const row_batch_start =
+        tile.batch_number == 0 ? 0 : batch_row_boundaries[tile.batch_number];
+
+    auto const participation_mask = __ballot_sync(0xFFFFFFFF, absolute_row < num_rows);
+
+    if (absolute_row < num_rows) {
+      auto const my_byte = input_data[row_offsets(absolute_row, row_batch_start) + validity_offset +
+                                      absolute_col / cols_per_read];
+
+      // so every thread that is participating in the warp has a byte, but it's row-based
+      // data and we need it in column-based. So we shuffle the bits around to make
+      // the bytes we actually write.
+      for (int i = 0, byte_mask = 0x1; i < cols_per_read && relative_col + i < num_columns;
```

Comment:
```suggestion
      for (int i = 0, byte_mask = 0x1; (i < cols_per_read) && ((relative_col + i) < num_columns);
```
Based on operator precedence and associativity, the original code does do the right thing, of course. I'd feel just a little safer if the ordering were explicit.